lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -0,0 +1,925 @@
+        )
+
+        try:
+            alert_rules_path = _resolve_dir_against_charm_path(charm, alert_rules_path)
+        except InvalidAlertRulePathError as e:
+            logger.debug(
+                "Invalid Prometheus alert rules folder at %s: %s",
+                e.alert_rules_absolute_path,
+                e.message,
+            )
+
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._alert_rules_path = alert_rules_path
+
+        self.topology = JujuTopology.from_charm(charm)
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_changed, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_departed, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_broken, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_joined, self._set_alerts_on_relation_changed)
+        self.framework.observe(self._charm.on.upgrade_charm, self._set_alerts_to_all_relation)
+
+    def _handle_endpoints_changed(self, event: RelationEvent):
+
+        self.on.endpoints_changed.emit(relation_id=event.relation.id)
+
+    def _set_alerts_on_relation_changed(self, event: RelationEvent):
+
+        self._set_alerts_to_relation(event.relation)
+
+    def _set_alerts_to_all_relation(self, _):
+        for relation in self.model.relations[self._relation_name]:
+            self._set_alerts_to_relation(relation)
+
+    def _set_alerts_to_relation(self, relation: Relation):
+        if not self._charm.unit.is_leader():
+            return
+
+        alert_rules = AlertRules(self.topology)
+        alert_rules.add_path(self._alert_rules_path, recursive=False)
+
+        alert_rules_as_dict = alert_rules.as_dict()
+
+        if alert_rules_as_dict:
+            relation.data[self._charm.app]["alert_rules"] = json.dumps(alert_rules_as_dict)
+
+    @property
+    def endpoints(self) -> List[Dict[str, str]]:
+        """A config object ready to be dropped in to a prometheus config file.
+
+        Returns:
+            A list of dictionaries where each dictionary provides information about
+            a single remote_write endpoint.
+        """
+        endpoints = []
+        for relation in self.model.relations[self._relation_name]:
+            for unit in relation.units:
+                if unit.app is self._charm.app:
+                    # This is a peer unit
+                    continue
+
+                if remote_write := relation.data[unit].get("remote_write"):
+                    deserialized_remote_write = json.loads(remote_write)
+                    endpoints.append(
+                        {
+                            "url": deserialized_remote_write["url"],
```

Comment:
Not sure if I get what you're wondering about, but if your question is why it's not just a list of url:s directly, it's because we might (will) want to add more properties to each entry down the road (think for instance TLS).